src/main/java/io/stargate/sgv2/jsonapi/service/operation/MetadataAttempt.java

```diff
@@ -0,0 +1,275 @@
+          return new ComplexTypes.MapType(
+              PrimitiveTypes.fromString(apiDataTypeDefKey.get().getApiType().getApiName()),
+              PrimitiveTypes.fromString(apiDataTypeDefValue.get().getApiType().getApiName()));
+        }
+      }
+      // return unsupported format
+      return new ComplexTypes.UnsupportedType(mt.asCql(true, false));
+
+    } else if (columnMetadata.getType()
+        instanceof com.datastax.oss.driver.api.core.type.ListType lt) {
+      if (!lt.isFrozen()) {
+        final Optional<ApiDataTypeDef> apiDataTypeDef = ApiDataTypeDefs.from(lt.getElementType());
+        if (apiDataTypeDef.isPresent()) {
+          return new ComplexTypes.ListType(
+              PrimitiveTypes.fromString(apiDataTypeDef.get().getApiType().getApiName()));
+        }
+      }
+      // return unsupported format
+      return new ComplexTypes.UnsupportedType(lt.asCql(true, false));
+
+    } else if (columnMetadata.getType()
+        instanceof com.datastax.oss.driver.api.core.type.SetType st) {
+      if (!st.isFrozen()) {
+        final Optional<ApiDataTypeDef> apiDataTypeDef = ApiDataTypeDefs.from(st.getElementType());
+        if (apiDataTypeDef.isPresent()) {
+          return new ComplexTypes.SetType(
+              PrimitiveTypes.fromString(apiDataTypeDef.get().getApiType().getApiName()));
+        }
+      }
+      // return unsupported format
+      return new ComplexTypes.UnsupportedType(st.asCql(true, false));
+    } else {
+      final Optional<ApiDataTypeDef> apiDataTypeDef =
+          ApiDataTypeDefs.from(columnMetadata.getType());
+      if (apiDataTypeDef.isPresent())
+        return PrimitiveTypes.fromString(apiDataTypeDef.get().getApiType().getApiName());
+      else {
+        // Need to return unsupported type
+        return new ComplexTypes.UnsupportedType(columnMetadata.getType().asCql(true, false));
+      }
+    }
+  }
+
+  /**
+   * Object used to build the response for listTables command
+   *
+   * @param name
+   * @param definition
+   */
+  @JsonPropertyOrder({"name", "definition"})
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  public record TableResponse(String name, TableDefinition definition) {
+
+    @JsonPropertyOrder({"columns", "primaryKey"})
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    record TableDefinition(Map<String, ColumnType> columns, PrimaryKey primaryKey) {}
+  }
+
+  protected List<TableSchemaObject> getTables() {
+    return keyspaceMetadata
+        .get()
+        // get all tables
+        .getTables()
+        .values()
+        .stream()
+        // filter for valid collections
+        .filter(TABLE_MATCHER.negate())
+        // map to name
+        .map(table -> TableSchemaObject.from(table, objectMapper))
+        // get as list
+        .toList();
+  }
+
+  /**
+   * EmptyAsyncResultSet implementation to be used only for metadata attempt where no cql query is
+   * run.
+   */
+  private static class EmptyAsyncResultSet implements AsyncResultSet {
```

Comment:
we can move this from an inner class, it's not related to the MetadataAttempt , the service.cqldriver package would be a good place